### DISCUSSION TITLE
Cruiser intents

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#25] [Q&A] [answered]
(continuing the discussion from the [old forum](https://groups.google.com/g/cruiser-users/c/JhGIK4nzk3A))

---

@blackboxembedded of WunderLINQ

Next Cruiser can parse OpenStreetMap directions urls and calculate routes.

Routing with 2 coordinates:
https://www.openstreetmap.org/directions?route=42.50646,1.521979;42.514612,1.526773

Routing from user location with 1 coordinate:
https://www.openstreetmap.org/directions?route=42.514612,1.526773

Can this work for you to call Cruiser for routing?
Do you need any other functionality for WunderLINQ?

**Comment by blackboxembedded** [comment]
These will be perfect, I'll update my end.

Thanks for adding them!

**Comment by devemux86** [comment]
@blackboxembedded the following should work:
```java
Intent intent = new Intent();
intent.setAction(Intent.ACTION_VIEW);
intent.setData(Uri.parse("https://www.openstreetmap.org/directions?route=42.50646,1.521979;42.514612,1.526773"));
intent.setPackage("gr.talent.cruiser");
startActivity(intent);
```

**Reply by blackboxembedded** [reply]
@devemux86  Sorry for the delay. Should 3.0.8 support it?  I tried with 3.0.8 and Cruiser isn't launching using the OpenStreetMap URL.

**Reply by devemux86** [reply]
@blackboxembedded It will be available in Cruiser 3.0.9.

**Reply by blackboxembedded** [reply]
Ah OK, sorry I will retest then.  Thanks again for adding this!

**Comment by devemux86** [comment]
@blackboxembedded 

I can include explicit Cruiser intents for map view, routing and navigation.

There is a discussion in the c:geo repository to call Cruiser with code samples: https://github.com/cgeo/cgeo/issues/13924#issuecomment-1413803745

If you have any comments, please let me know!

**Reply by blackboxembedded** [reply]
These new intents would work for me as well.  I noticed you had two examples, a 'Routing' and 'Navigation' intent, what is the difference in Cruiser?

**Reply by devemux86** [reply]
I am still testing if the parameters have the best structure, especially for the multiple coordinates.

>  I noticed you had two examples, a 'Routing' and 'Navigation' intent, what is the difference in Cruiser?

Navigation intent performs a routing + enables the navigation.

If possible,
better use the intent with the 2 coordinates and pass user location + destination.

When first start Cruiser (or any Android app), it takes time to find user location.
If it is already known externally, better pass location as start in routing request.

**Reply by blackboxembedded** [reply]
Thanks for the explanation.  I'm able to pass the current location and the destination location.

**Reply by devemux86** [reply]
I will inform when the structure of intents is final.
They will be included in the next Cruiser 3.0.9.

**Comment by devemux86** [comment, marked as answer]
@blackboxembedded 

Please see below the Cruiser intents, will be available in Cruiser 3.0.9.

For routing and navigation, can pass an array of coordinates and better _set user location as the start_.
The naming of the waypoints is optional, if one or less names are passed, only destination is named.

**Map view**:
```java
Intent intent = new Intent();
intent.setAction("com.devemux86.intent.action.MAP_VIEW");
intent.setPackage("gr.talent.cruiser");
intent.putExtra("LATITUDE", 42.514612);
intent.putExtra("LONGITUDE", 1.526773);
intent.putExtra("NAME", "Test");
startActivity(intent);
```

**Routing**:
```java
Intent intent = new Intent();
intent.setAction("com.devemux86.intent.action.ROUTING");
intent.setPackage("gr.talent.cruiser");
intent.putExtra("LATITUDE", new double[]{42.50646, 42.508654, 42.514612});
intent.putExtra("LONGITUDE", new double[]{1.521979, 1.529916, 1.526773});
intent.putExtra("NAME", new String[]{"Beginning", "Stop!", "Destination"});
startActivity(intent);
```

**Navigation**:
```java
Intent intent = new Intent();
intent.setAction("com.devemux86.intent.action.NAVIGATION");
intent.setPackage("gr.talent.cruiser");
intent.putExtra("LATITUDE", new double[]{42.50646, 42.508654, 42.514612});
intent.putExtra("LONGITUDE", new double[]{1.521979, 1.529916, 1.526773});
intent.putExtra("NAME", new String[]{"Beginning", "Stop!", "Destination"});
startActivity(intent);
```

**Reply by blackboxembedded** [reply]
Thanks for the update.  I will update my app to be ready for 3.0.9.

**Reply by devemux86** [reply]
@blackboxembedded 

I updated above the examples in https://github.com/devemux86/cruiser/discussions/25#discussioncomment-4880421 with minor improvement in action names.

**Comment by devemux86** [comment]
@blackboxembedded 

[Cruiser 3.0.9](https://play.google.com/store/apps/details?id=gr.talent.cruiser) is available with the new intents mentioned above in https://github.com/devemux86/cruiser/discussions/25#discussioncomment-4880421.

**Reply by blackboxembedded** [reply]
Sorry for the delay. I did some testing this morning and I'm unable to get Cruiser to launch with any of the new intents.  I even tried embedding your examples directly.  I had a similar problem in an app when I had forgotten to set android:exported="true"

**Reply by devemux86** [reply]
@blackboxembedded 

Sorry for that! It was a problem with the renaming of actions.
It will work properly in the next Cruiser 3.0.10 (coming soon).

**Comment by devemux86** [comment]
@blackboxembedded 

[Cruiser intents](https://github.com/devemux86/cruiser/discussions/25#discussioncomment-4880421) should work properly with [Cruiser 3.0.10](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by blackboxembedded** [reply]
Everything appears to be working as expected!  Thanks for your work on this.

**Comment by devemux86** [comment]
I missed WunderLINQ news about Cruiser integration, thanks @blackboxembedded!
- https://blackboxembedded.com/blogs/news/cruiser
- https://www.youtube.com/watch?v=NlWjEgTcSKs

**Comment by blackboxembedded** [comment]
@devemux86 

Is there a way through intents to have Cruiser navigate to the nearest fuel station?

**Reply by devemux86** [reply]
@blackboxembedded 

I can create something like this, we just need to design how it could work. :slightly_smiling_face: 

Do you mean when navigation is active and nearby POI (fuel stations) are enabled / displayed
to reroute to the nearest / next fuel station?

This is already possible in UI when we press a displayed fuel station panel:
- https://github.com/devemux86/cruiser/discussions/9#discussioncomment-6550514

Or when navigation is active, regardless if nearby POI (fuel stations) are enabled / displayed or not,
to be able to search offline for the nearest / next fuel station and then reroute to it?

**Reply by blackboxembedded** [reply]
@devemux86 Thanks for getting back to me.

The WunderLINQ monitors the motorcycle's fuel level and when low can prompt the user to route to the nearest fuel station.  The WunderLINQ app could send an intent to tell Cruiser to add the nearest fuel station to their existing navigation or as a new target.

**Reply by devemux86** [reply]
> The WunderLINQ app could send an intent to tell Cruiser to add the nearest fuel station to their existing navigation or as a new target.

@blackboxembedded in order to not clear user's active route and navigation,
better to add the nearest fuel station to the existing route as next waypoint,
merged with the rest of the route that remains unchanged?

What you describe is an intent to do what UI is already doing?
- https://github.com/devemux86/cruiser/discussions/9#discussioncomment-6550514

Do I understand correctly the workflow?

**Reply by WolfgangZ42** [reply]
I don't need software that alerts me when my fuel is empty. And I rarely need to know where the nearest gas station is either.

**Reply by MotoUKRider** [reply]
Hands-up if you remember having to unlock the fuel tank, rock the bike side-to-side and guesstimate how much you had left before hitting the reserve tap?!

**Reply by WolfgangZ42** [reply]
Old enough am I for this stuff - but I didn't get my motorcycle license until I was 50.

**Comment by blackboxembedded** [comment]
> > The WunderLINQ app could send an intent to tell Cruiser to add the nearest fuel station to their existing navigation or as a new target.
> 
> @blackboxembedded in order to not clear user's active route and navigation, better to add the nearest fuel station to the existing route as next waypoint, merged with the rest of the route that remains unchanged?
> 
> What you describe is an intent to do what UI is already doing?
> 
> * [Nearby POI in navigation #9 (comment)](https://github.com/devemux86/cruiser/discussions/9#discussioncomment-6550514)
> 
> Do I understand correctly the workflow?

@devemux86 You are exactly correct on the desired behavior. :)

**Reply by devemux86** [reply]
@blackboxembedded thanks, I will check it. :slightly_smiling_face:

**Comment by devemux86** [comment]
> The WunderLINQ app could send an intent to tell Cruiser to add the nearest fuel station to their existing navigation or as a new target.

@blackboxembedded please test Cruiser 3.0.36.1 from [here](https://www.dropbox.com/sh/itq3e3d0wr0zp3q/AADZ0BYypUf91GhMVKryZzmba) using the following intent.

Navigation will then go from image 1 after rerouting to image 2. Is this what you mean?
(nearby fuel stations must be enabled in navigation settings)

```java
Intent intent = new Intent();
intent.setAction("com.devemux86.intent.action.NAVIGATION");
intent.setPackage("gr.talent.cruiser");
intent.putExtra("FUEL", true);
startActivity(intent);
```

<img src="https://github.com/devemux86/cruiser/assets/3484020/fe89aae8-1b2e-4e4e-97c1-2dcf44ba819e" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/5fec0d8e-5332-42a0-a12a-9aef766b1fd1" width="50%"/>

**Reply by blackboxembedded** [reply]
@devemux86 Thanks, that looks perfect. I will test and report back.

**Reply by blackboxembedded** [reply]
@devemux86 I was unable to get it to work, it opens Cruiser but doesn't add a fuel stop.

**Reply by devemux86** [reply]
@blackboxembedded how do you test it?

There are some requirements:
- Nearby fuel stations are enabled in navigation settings
- Navigation is already running, displaying fuel stations

Then instead of long press on fuel panel and select rerouting, we do the same via an intent.
So it is the existing function, but activated differently.

This is what we discussed, do you mean something else?

**Reply by blackboxembedded** [reply]
@devemux86 

Sorry, I missed the bit about having to turn on the display of fuel stations, that is most likely the problem. I've looked through Settings -> Navigation and elsewhere and I can't seem to find the option for this. Can you point me to the correct spot?

**Reply by devemux86** [reply]
@blackboxembedded 

It is in "Settings | Navigation | Display | Nearby points of interest" or search settings for "fuel".

<img src="https://github.com/devemux86/cruiser/assets/3484020/ce85a0e8-6124-4487-9b65-e6c928ade796" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/7e4188b5-40ec-4756-a053-910b516e255e" width="50%"/>
<img src="https://github.com/devemux86/cruiser/assets/3484020/f4c2c78f-d182-4e9c-9fea-33bd45908424" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/24a071ae-5be4-417d-8e0f-c70e7644da30" width="50%"/>

**Reply by blackboxembedded** [reply]
@devemux86 Sorry for the delay.  I can confirm it works!  Thanks for adding it so quickly. I had an idea as I was testing on the road. 
 Would it be possible for that intent to start navigation to the nearest fuel station if not already navigating?

**Reply by devemux86** [reply]
@blackboxembedded thanks for the tests!

> Would it be possible for that intent to start navigation to the nearest fuel station if not already navigating?

This is more complicated as it involves more steps and several variables:
- If we have an existing route, do we clear it first to calculate a new one?
- A search for the nearest fuel station must first be completed
- Then we call route calculation from user location to the fuel station
- If above steps finish successfully, navigation can start with new route

Do you mean something like this?

**Reply by blackboxembedded** [reply]
@devemux86 

Sorry, when I said navigation I meant route.

The above is correct except for the first point.  The current behavior is perfect when there is an active route, this would be if Cruiser receives the intent and there is no active route. 

If there is no existing route, then do the following:

- A search for the nearest fuel station must first be completed
- Then we call route calculation from user location to the fuel station
- If above steps finish successfully, navigation can start with new route

**Reply by devemux86** [reply]
@blackboxembedded 

> The current behavior is perfect when there is an active route

The current intent works with active navigation and nearby fuel stations.
(so we also have a route)

The other cases can be:
- Navigation is off
  - Route exists
  - Route not exists

**Reply by blackboxembedded** [reply]
@devemux86 

Ah, OK I follow the convention.

So I think it would be OK to clear the route if Navigation is off.

**Reply by devemux86** [reply]
> Would it be possible for that intent to start navigation to the nearest fuel station if not already navigating?

@blackboxembedded please try the latest:
- https://github.com/devemux86/cruiser/discussions/390

I summarize again the workflow in case we need something else:
- Navigation **On** / **Paused**: rerouting to the nearest fuel station, _route remains the same_

- Navigation **Off**: a **new** route is calculated to the nearest fuel station + start navigation

What search radius (now 20 km) around user location should we use?

**Reply by blackboxembedded** [reply]
@devemux86 

I can confirm both scenarios work for me as described. 

For the search, the radius may be just a bit bigger, is 30 km OK?

**Reply by devemux86** [reply]
@blackboxembedded thanks for the feedback!

I will increase the distance to 30 km.

**Comment by devemux86** [comment]
> Is there a way through intents to have Cruiser navigate to the nearest fuel station?

@blackboxembedded it is now available in [Cruiser 3.0.37](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by blackboxembedded** [reply]
@devemux86 Thanks for the heads up!  Happy Holidays!
